[Cache] Add nonce based cache invalidation to ApcuAdapter

nicolas-grekas opened the pull request:
| Q | A |
| --- | --- |
| Branch? | 3.1 |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

APCu cache invalidation is tedious because it can't be done from the cli.
This PR adds and wires a nonce based cache invalidation strategy to work around this limitation.

nicolas-grekas committed:
[Cache] Add nonce based cache invalidation to ApcuAdapter

xabbuh commented on src/Symfony/Component/Cache/Adapter/ApcuAdapter.php:
```diff
@@ -27,22 +30,49 @@ public function __construct($namespace = '', $defaultLifetime = 0)
+                }
+            }
+            foreach ($values as $id => $v) {
+                $values[$id] = unserialize(substr($v, $this->nonceLen));
```
Why not moving this to the loop above?

nicolas-grekas replied:
Good question: because unserialization involves userland code, which means can be slow. But IMHO we need to invalidate fast.

xabbuh commented:
Having the nonce as a container parameter makes the APCu cache barely unusable in the `dev` environment where the container may be rebuilt quite often, doesn't it?

nicolas-grekas commented:
Except that caches are not enabled in debug mode

nicolas-grekas commented:
I simplified the implementation to the very minimum so that no overhead is added anymore to each apcu_fetch/store. I also replaced the container parameter by an inline value for the nonce.

It has the drawback that @xabbuh outlined: the cache **in dev** is going to be cleared more often than strictly required. Yet I think this does not matter compared to the main benefit (we even already disable most caches in dev):

As you know, there is currently no way to invalidate an FPM/Apache apcu cache from the CLI, other than restarting FPM/Apache on deployment. Even if it's a best practice, not everyone is able to do it. 
Nonce-based invalidation fixes this. Thus, it makes apcu usable again **in prod** for everyone, by default.

dunglas commented:
👍

nicolas-grekas commented:
ping @symfony/deciders votes pendind for 3.1-beta which should/may be released tomorrow

fabpot commented:
Thank you @nicolas-grekas.
